File: src/data/courses/dsa/dynamic-programming/lis-sequence/number-of-lis/code/java/NumberOfLis.java
class Solution {
    public int findNumberOfLIS(int[] nums) {
        int n=nums.length,maxLen=1; int[] dp=new int[n],cnt=new int[n];
        java.util.Arrays.fill(dp,1); java.util.Arrays.fill(cnt,1);
        for(int i=1;i<n;i++) for(int j=0;j<i;j++) if(nums[j]<nums[i]){
            if(dp[j]+1>dp[i]){dp[i]=dp[j]+1;cnt[i]=cnt[j];}
            else if(dp[j]+1==dp[i]) cnt[i]+=cnt[j];}
        for(int d:dp) maxLen=Math.max(maxLen,d);
        int t=0; for(int i=0;i<n;i++) if(dp[i]==maxLen) t+=cnt[i]; return t;
    }
}

File: src/data/courses/dsa/dynamic-programming/lis-sequence/number-of-lis/code/python/number_of_lis.py
class Solution:
    def findNumberOfLIS(self, nums) -> int:
        n=len(nums); dp=[1]*n; cnt=[1]*n
        for i in range(1,n):
            for j in range(i):
                if nums[j]<nums[i]:
                    if dp[j]+1>dp[i]: dp[i]=dp[j]+1; cnt[i]=cnt[j]
                    elif dp[j]+1==dp[i]: cnt[i]+=cnt[j]
        max_len=max(dp)
        return sum(cnt[i] for i in range(n) if dp[i]==max_len)
